File: lab9-structural-patterns-CichyJohnny/Facade.java
class FlightSearch {
    public void search() {
        System.out.println("Searching for flights from New York to Warsaw");
        System.out.println("Flights found:\n\t- 12:00\n\t- 15:00\n\t- 18:00");
    }
}

class BookingSystem {
    public void book() {
        System.out.println("Booking flight for Jan Kowalski");
    }

}

class PaymentProcessor {
    public void pay() {
        System.out.println("Processing payment of 250.00 USD");
    }
   
}

class FlightFacade {
    private final FlightSearch flightSearch;
    private final BookingSystem bookingSystem;
    private final PaymentProcessor paymentProcessor;

    public FlightFacade() {
        flightSearch = new FlightSearch();
        bookingSystem = new BookingSystem();
        paymentProcessor = new PaymentProcessor();
    }

    public void bookFlight() {
        flightSearch.search();
        bookingSystem.book();
        paymentProcessor.pay();

        System.out.println("Flight booked.");
    }
    
}

public class Facade {
    public static void main(String[] args) {
        FlightFacade facade = new FlightFacade();
        
        facade.bookFlight();
    }
}

File: lab9-structural-patterns-CichyJohnny/Facade.cpp
#include <iostream>

class FlightSearch {
public:
    void search() {
        std::cout << "Searching for flights from New York to Warsaw" << std::endl;
        std::cout << "Flights found:\n\t- 12:00\n\t- 15:00\n\t- 18:00" << std::endl;
    }
};

class BookingSystem {
public:
    void book() {
        std::cout << "Booking flight for Jan Kowalski" << std::endl;
    }
};

class PaymentProcessor {
public:
    void pay() {
        std::cout << "Processing payment of 250.00 USD" << std::endl;
    }
};

class FlightFacade {
private:
    FlightSearch flightSearch;
    BookingSystem bookingSystem;
    PaymentProcessor paymentProcessor;

public:
    FlightFacade() {
        flightSearch = FlightSearch();
        bookingSystem = BookingSystem();
        paymentProcessor = PaymentProcessor();
    }

    void bookFlight() {
        flightSearch.search();
        bookingSystem.book();
        paymentProcessor.pay();
    }
};

int main() {
    FlightFacade fs = FlightFacade();

    fs.bookFlight();

    return 0;
}